### ISSUE TITLE
General inference executor for any Terratorch model available in HuggingFace

### ISSUE THREAD

**Issue by mgazz** [terrastackai/terratorch#788]
**Is your feature request related to a problem? Please describe.**
The current Terratorch/vLLM integration initialises Terratorch models using task classes but the input configuration is specific for the SemanticSegmentationTask.

**Describe the solution you'd like**
We would like to generalise the loading of Terratorch models in vLLM. To do this we need a class: 
- takes as input a configuration in the form of a dictionary
- supports the initialisation of any Terratorch model currently available in HF. 

**Additional context (optional)**
* The structure of the input config should be the same no matter the model.
* We want to initialise models purely for inference. No training information should be required to load models.

@romeokienzler @paolofraccaro

**Comment by Joao-L-S-Almeida** [comment]
This is a suggestion which follows the structure of the tasks currently avilable on terratorch:
```
from collections.abc import Sequence
from functools import partial
from typing import Any

import logging
import lightning
import matplotlib.pyplot as plt
import torch
import torch.nn.functional as F
from lightning.pytorch.callbacks import Callback
from torch import Tensor, nn
from torchgeo.datasets.utils import unbind_samples
from torchmetrics import MeanAbsoluteError, MeanSquaredError, MetricCollection, R2Score
from torchmetrics.metric import Metric
from torchmetrics.wrappers.abstract import WrapperMetric

from terratorch.models.model import AuxiliaryHead, Model, ModelOutput
from terratorch.registry.registry import MODEL_FACTORY_REGISTRY
from terratorch.tasks.loss_handler import LossHandler
from terratorch.tasks.optimizer_factory import optimizer_factory
from terratorch.tasks.tiled_inference import TiledInferenceParameters, tiled_inference
from terratorch.tasks.base_task import TerraTorchTask

class InferenceTask(TerraTorchTask):
    """Pixelwise Regression Task that accepts models from a range of sources.

    This class is analog in functionality to PixelwiseRegressionTask defined by torchgeo.
    However, it has some important differences:
        - Accepts the specification of a model factory
        - Logs metrics per class
        - Does not have any callbacks by default (TorchGeo tasks do early stopping by default)
        - Allows the setting of optimizers in the constructor
        - Allows to evaluate on multiple test dataloaders"""

    def __init__(
        self,
        model_args: dict | None = None,
        model_factory: str | None = None,
        model: torch.nn.Module | None = None,
        freeze_backbone: bool = True,
        freeze_decoder: bool = True,
        tiled_inference_parameters: dict | None = None,
    ) -> None:

        """Constructor

        Args:
            model_args (Dict): Arguments passed to the model factory.
            model_factory (str, optional): Name of ModelFactory class to be used to instantiate the model.
                Is ignored when model is provided.
            model (torch.nn.Module, optional): Custom model.
            tiled_inference_parameters (dict | None, optional): Inference parameters
        """

        self.model_args = model_args
        self.tiled_inference_parameters = tiled_inference_parameters

        if model is not None and model_factory is not None:
            logger.warning("A model_factory and a model was provided. The model_factory is ignored.")
        if model is None and model_factory is None:
            raise ValueError("A model_factory or a model (torch.nn.Module) must be provided.")

        if model_factory and model is None:
            self.model_factory = MODEL_FACTORY_REGISTRY.build(model_factory)

        super().__init__(
            task="inference", 
            )

        if model:
            # Custom_model
            self.model = model

    def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> Tensor:
        """Compute the predicted class probabilities.

        Args:
            batch: The output of your DataLoader.
            batch_idx: Integer displaying index of this batch.
            dataloader_idx: Index of the current dataloader.

        Returns:
            Output predicted probabilities.
        """

        def model_forward(x, **kwargs):
            return self(x)

        if self.tiled_inference_parameters:
            y_hat: Tensor = tiled_inference(model_forward, batch, **self.tiled_inference_parameters, **rest)
        else:
            y_hat: Tensor = self(batch)

        return y_hat
```
It will basically get a datamodule and a backbone instance and perform inferences. It also could be defined from a YAML file.

**Comment by mgazz** [comment]
@Joao-L-S-Almeida thanks a million for the sample. That is helpful. I started working on a prototype that uses the same approach for initalising models using factories. 

I think our use-case is even simpler because the tiling is done outside so we don't have to worry about it. 

Currently I am facing two challenges:

### Challenge 1: different signatures for the forward method

In vllm the input arrives in the form of `**kwargs` so (key,value).

For Prithvi the model is a `PixelWiseModel` and takes as input a Tensor and (key,value) arguments. [ref](https://github.com/IBM/terratorch/blob/main/terratorch/models/pixel_wise_model.py#L100)

```python
def forward(self, x: torch.Tensor, **kwargs) -> ModelOutput:
```

For granite-geospatial-wxc-downscaling instead we initialise a `ClimateDownscaleFinetuneModel` model coming from the `granitewxc` library and the forward method takes a dict  ([ref](https://github.com/IBM/granite-wxc/blob/main/granitewxc/models/finetune_model.py#L248)

```
    def forward(self, batch: dict[str, torch.tensor]):
```

This makes it hard to generalise. 

*Potential solution:* standardise the input to the forward method. `batch: dict[str, torch.tensor]` would be preferable because it fits the way vllm passes data to model executor. 

### Challenge 2: Provide valid input for the vLLM warmup

vLLM requires a synthetic input sample to test that there is enough memory to run the model.

For Prithvi the input is a `torch.Tensor` 

while for [granite-geospatial-wxc-downscaling](https://huggingface.co/ibm-granite/granite-geospatial-wxc-downscaling) is a dict of tensors. This is based on the sample extracted from the `Merra2DownscaleNonGeoDataModule` described in the [WxCTutorialDownscaling tutorial](https://github.com/IBM/terratorch/blob/main/examples/notebooks/WxCTutorialDownscaling.ipynb)
```

            "y": Tensor,
            "static_y": Tensor
            "climate_y": Tensor
            "x":Tensor
            "static_x": Tensor
            "climate_x": Tensor

```

The tricky part is that we need to know both the name and the shape of the tensor. 


*Potential solution:* Add a section in the config where we specify the input and tensor shape. For instance we could have 

The section could have the following structure:
```yaml
input:
  data:
     key1:
       type: tensor
       shape: []
     key2:
       type: tensor
       shape: []
  target: key of one of input tensors. (optional)
```

Defining `target`, we specify which tensor should be passed to the `forward()` method. When not specified we pass the whole data dict.

For Prithvi it would become as following. with `targe: pixel_values` we know which tensor we should pass as input, the rests are additional arguments. 
```yaml
input:
  data:
    pixel_values: 
       type: tensor
       shape: []
    location_coords: 
       type: tensor
       shape: []
    target: pixel_values
```

For [granite-geospatial-wxc-downscaling] we don't specify the target so the data dict is passed as is.
```yaml
input:
  data:
    - x: 
        type: tensor
        shape: []
    - y: 
        type: tensor
        shape: []
....
```

*Note:* this way we do not need to change the signature of the `forward` methods because the `input` section of the config tells us what to use. 

That said, I do not have enough awareness on Terratorch to say if this approach is generalisable to other models.

FYI: @christian-pinto @paolofraccaro @romeokienzler

**Comment by mgazz** [comment]
For reference, this is my attempt to create a generalised inference runner: https://github.com/IBM/terratorch/blob/8c4067fd7e8325ab8ea10480b9038c430f6306e3/terratorch/vllm.py

As you can see, not the [dummy data](https://github.com/IBM/terratorch/blob/8c4067fd7e8325ab8ea10480b9038c430f6306e3/terratorch/vllm.py#L37) and [data fields](https://github.com/IBM/terratorch/blob/8c4067fd7e8325ab8ea10480b9038c430f6306e3/terratorch/vllm.py#L15) are generated based on the task. This is a big assumption that implies that all models part of the same task take the same input. An input section would allow us to dynamically generate synthetic data and fields.

**Comment by mgazz** [comment]
Looking at this with @christian-pinto, we noticed that Prithvi specifies the tensor in input as `x`. This means that we could pass the input as kwargs with the target tensor with key "x". The problem of knowing the right key strings still stands.

**Comment by mgazz** [comment]
I've built a PoC showing how this would work. 

We define the input to the model as [InputDefinition](https://github.com/mgazz/terratorch/blob/vllm/terratorch/vllm.py#L20) and this is used to
- [create the dummy data for the warmup](https://github.com/mgazz/terratorch/blob/vllm/terratorch/vllm.py#L50)
- [process and validate the input](https://github.com/mgazz/terratorch/blob/vllm/terratorch/vllm.py#L81)
- [understand how to pass data to the model ](https://github.com/mgazz/terratorch/blob/vllm/terratorch/vllm.py#L97)

Here are examples of `config.json` to run the models in vLLM. In `pretrained_cfg.input` we specify the structure of the input data. I will build huggingface repositories as examples. 

<details>

<summary>Prithvi-EO-V2-300M-TL-Sen1Floods11 (expand to see the content)</summary>
```json
{
  "architectures": [
    "PrithviGeoSpatialMAE"
  ],
  "num_classes": 0,
  "pretrained_cfg": {
    "input":{
      "target": "pixel_values",
      "data":{
        "pixel_values":{
          "type": "torch.Tensor",
          "shape": [6, 512, 512]
        },
        "location_coords":{
          "type":"torch.Tensor",
          "shape": [1, 2]
        }
      }
    },
    "seed_everything": 0,
    "trainer": {
      "accelerator": "auto",
      "strategy": "auto",
      "devices": "auto",
      "num_nodes": 1,
      "precision": "16-mixed",
      "logger": true,
      "callbacks": [
        {
          "class_path": "lightning.pytorch.callbacks.RichProgressBar",
          "init_args": {
            "refresh_rate": 1,
            "leave": false,
            "theme": {
              "description": "white",
              "progress_bar": "#6206E0",
              "progress_bar_finished": "#6206E0",
              "progress_bar_pulse": "#6206E0",
              "batch_progress": "white",
              "time": "grey54",
              "processing_speed": "grey70",
              "metrics": "white",
              "metrics_text_delimiter": " ",
              "metrics_format": ".3f"
            }
          }
        },
        {
          "class_path": "lightning.pytorch.callbacks.LearningRateMonitor",
          "init_args": {
            "logging_interval": "epoch",
            "log_momentum": false,
            "log_weight_decay": false
          }
        },
        {
          "class_path": "lightning.pytorch.callbacks.EarlyStopping",
          "init_args": {
            "monitor": "val/loss",
            "min_delta": 0.0,
            "patience": 20,
            "verbose": false,
            "mode": "min",
            "strict": true,
            "check_finite": true,
            "log_rank_zero_only": false
          }
        }
      ],
      "fast_dev_run": false,
      "max_epochs": 50,
      "max_steps": -1,
      "overfit_batches": 0.0,
      "check_val_every_n_epoch": 2,
      "log_every_n_steps": 10,
      "enable_checkpointing": true,
      "accumulate_grad_batches": 1,
      "inference_mode": true,
      "use_distributed_sampler": true,
      "detect_anomaly": false,
      "barebones": false,
      "sync_batchnorm": false,
      "reload_dataloaders_every_n_epochs": 0,
      "default_root_dir": "/dccstor/geofm-finetuning/benchmark-geo-bench-paolo/"
    },
    "model": {
      "class_path": "terratorch.tasks.SemanticSegmentationTask",
      "init_args": {
        "model_args": {
          "backbone_pretrained": true,
          "backbone": "prithvi_eo_v2_300_tl",
          "decoder": "UperNetDecoder",
          "decoder_channels": 256,
          "decoder_scale_modules": true,
          "num_classes": 2,
          "rescale": true,
          "backbone_bands": [
            "BLUE",
            "GREEN",
            "RED",
            "NIR_NARROW",
            "SWIR_1",
            "SWIR_2"
          ],
          "head_dropout": 0.1,
          "necks": [
            {
              "name": "SelectIndices",
              "indices": [
                5,
                11,
                17,
                23
              ]
            },
            {
              "name": "ReshapeTokensToImage"
            }
          ]
        },
        "model_factory": "EncoderDecoderFactory",
        "loss": "ce",
        "ignore_index": -1,
        "lr": 0.001,
        "freeze_backbone": false,
        "freeze_decoder": false,
        "plot_on_val": 10
      }
    },
    "data": {
      "class_path": "terratorch.datamodules.Sen1Floods11NonGeoDataModule",
      "init_args": {
        "data_root": "/dccstor/geofm-finetuning/datasets/sen1floods11",
        "batch_size": 16,
        "num_workers": 8,
        "bands": [
          "BLUE",
          "GREEN",
          "RED",
          "NIR_NARROW",
          "SWIR_1",
          "SWIR_2"
        ],
        "train_transform": [
          {
            "class_path": "albumentations.Resize",
            "init_args": {
              "height": 448,
              "width": 448,
              "interpolation": 1,
              "always_apply": false,
              "p": 1
            }
          },
          {
            "class_path": "albumentations.RandomCrop",
            "init_args": {
              "height": 224,
              "width": 224,
              "always_apply": false,
              "p": 1.0
            }
          },
          {
            "class_path": "albumentations.HorizontalFlip",
            "init_args": {
              "always_apply": false,
              "p": 0.5
            }
          },
          {
            "class_path": "albumentations.VerticalFlip",
            "init_args": {
              "always_apply": false,
              "p": 0.5
            }
          },
          {
            "class_path": "albumentations.pytorch.ToTensorV2",
            "init_args": {
              "transpose_mask": false,
              "always_apply": true,
              "p": 1.0
            }
          }
        ],
        "val_transform": [
          {
            "class_path": "albumentations.Resize",
            "init_args": {
              "height": 448,
              "width": 448,
              "interpolation": 1,
              "always_apply": false,
              "p": 1
            }
          },
          {
            "class_path": "albumentations.pytorch.ToTensorV2",
            "init_args": {
              "transpose_mask": false,
              "always_apply": true,
              "p": 1.0
            }
          }
        ],
        "test_transform": [
          {
            "class_path": "albumentations.Resize",
            "init_args": {
              "height": 448,
              "width": 448,
              "interpolation": 1,
              "always_apply": false,
              "p": 1
            }
          },
          {
            "class_path": "albumentations.pytorch.ToTensorV2",
            "init_args": {
              "transpose_mask": false,
              "always_apply": true,
              "p": 1.0
            }
          }
        ],
        "drop_last": true,
        "constant_scale": 0.0001,
        "no_data_replace": 0.0,
        "no_label_replace": -1,
        "use_metadata": false
      }
    },
    "out_dtype": "int16",
    "deploy_config_file": true,
    "optimizer": {
      "class_path": "torch.optim.AdamW",
      "init_args": {
        "lr": 5e-05,
        "betas": [
          0.9,
          0.999
        ],
        "eps": 1e-08,
        "weight_decay": 0.05,
        "amsgrad": false,
        "maximize": false,
        "capturable": false,
        "differentiable": false
      }
    },
    "lr_scheduler": {
      "class_path": "torch.optim.lr_scheduler.CosineAnnealingLR",
      "init_args": {
        "T_max": 50,
        "eta_min": 0,
        "last_epoch": -1,
        "verbose": "deprecated"
      }
    }
  }
}
```
</details>

<details>

<summary>granite-geospatial-wxc-downscaling (expand to see the content) </summary>
```json
{
  "architectures": [
    "PrithviGeoSpatialMAE"
  ],
  "num_classes": 0,
  "pretrained_cfg": {
	"input": {
		"data":{
			"x":{
				"type": "torch.Tensor",
				"shape": [280, 60, 96]

			},
			"static_x":{
				"type": "torch.Tensor",
				"shape": [11, 60, 96]

			},
			"climate_x":{
				"type": "torch.Tensor",
				"shape": [140, 60, 96]

			},
			"y":{
				"type": "torch.Tensor",
				"shape": [1, 360, 576]

			},
			"static_y":{
				"type": "torch.Tensor",
				"shape": [11, 360, 576]

			},
			"climate_y":{
				"type": "torch.Tensor",
				"shape": [1, 360, 576]

			}

		}

	},
    "model": {
      "class_path": "WxCDownscalingTask",
      "init_args":{
	"model_factory": "WxCModelFactory",
        "model_args":{
		"backbone":  "prithviwxc",
		"aux_decoders": "downscaler",
		"in_channels": 1280,
		"input_size_time": 1,
		"n_lats_px": 64,
		"n_lons_px": 128,
		"in_channels_static": 3,
		"input_scalers_epsilon": 0,
		"static_input_scalers_epsilon": 0,
		"patch_size_px": [2, 2],
		"mask_unit_size_px": [8, 16],
		"mask_ratio_inputs": 0.5,
		"embed_dim": 2560,
		"n_blocks_encoder": 12,
		"n_blocks_decoder": 2,
		"mlp_multiplier": 4,
		"n_heads": 16,
		"dropout": 0.0,
		"drop_path": 0.05,
		"parameter_dropout": 0.0,
		"residual": "none",
		"masking_mode": "both",
		"positional_encoding": "absolute",
		"config_path": "/workspace/geospatial_vllm_test_scripts/models/granite-geospatial-wxc-downscaling/config.yaml",
		"wxc_auxiliary_data_path": "/workspace/geospatial_vllm_test_scripts/models/granite-geospatial-wxc-downscaling"
	}
      }
    }
  }

}

```
</details>

**Comment by mgazz** [comment]
@Joao-L-S-Almeida @romeokienzler @christian-pinto @paolofraccaro please let me know if you have feedbacks on this approach. The implication is that models to be used in vLLM will need an `input` section in the `config.json` explaining the structure of the data expected by the forward method.

**Comment by romeokienzler** [comment]
Dear @mgazz so sorry, this slipped through somehow

Spoiler: I love your suggestion, but I think the might break quite some code

>signatures for the forward method
This is something which I would really like to see.  Also for the return value.
The CV models  support only a single tensor whereas the climate models need at least 4 as input. So I'd say we could agree on a  generic signature covering those cases and potential future cases and create a wrapper. For the old code we map the wrapper's signature to the underlying model, for new models we implement according to the new signature.

I fear we would also need to change/wrap all data loaders

So it sound more like a TerraTorch 3.0 thing to me atm.


> Provide valid input for the vLLM warmup
I think this is much more feasible. Any exporter needs it (ONNX, TensorRT, NorthPole, ...)

So happy to discuss offline or in the TerraTorch community call
